text/0000-promise.md

```diff
@@ -0,0 +1,260 @@
+
+    setTimeout(() => d.resolve(), 5000);
+
+    return promise;
+}
+```
+
+This could be written in Neon with the following:
+
+```rust
+fn async_method(cx: FunctionContext) -> JsResult<JsPromise> {
+    let queue = cx.queue();
+    let (promise, d) = cx.promise();
+
+    std::thread::spawn(move || {
+        std::thread::sleep(std::time::Duration::from_millis(5000));
+
+        queue.send(|cx| d.resolve(cx.undefined()));
+    });
+
+    Ok(promise)
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `JsPromise` API consists of two structs:
+
+* `JsPromise`. Opaque value; only useful for passing back to JavaScript.
+* `Deferred`. Handle for resolving and rejecting the related `Promise`.
+
+They may only be created with the `deferred` method on `Context`.
+
+```rust
+trait Context {
+    fn deferred(&mut self) -> (Handle<JsPromise>, Deferred);
+}
+```
+
+## `JsPromise`
+
+Opaque handle that represents a JavaScript `Promise`. It is not threadsafe.
+
+```rust
+/// A JavaScript `Promise`
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct JsPromise(raw::Local);
+
+impl Value for JsPromise {}
+
+impl Managed for JsPromise {
+    fn to_raw(self) -> raw::Local { self.0 }
+    fn from_raw(h: raw::Local) -> Self { JsPromise(h) }
+}
+
+impl ValueInternal for JsPromise {
+    fn name() -> String { "promise".to_string() }
+    fn is_typeof<Other: Value>(env: Env, other: Other) -> bool {
+        unsafe { neon_runtime::tag::is_promise(env.to_raw(), other.to_raw()) }
+    }
+}
+
+unsafe impl This for JsPromise {
+    fn as_this(_env: Env, h: raw::Local) -> Self {
+        JsPromise(h)
+    }
+}
+
+impl JsPromise {
+    pub(crate) fn new_internal<'a>(value: raw::Local) -> Handle<'a, JsPromise> {
+        Handle::new_internal(JsPromise(value))
+    }
+}
+```
+
+## `Deferred`
+
+`Send` handle for resolving or rejecting a `JsPromise`.
+
+```rust
+pub struct Deferred(*mut c_void);
+
+unsafe impl Send for Deferred {}
+
+impl Deferred {
+    /// Resolves a deferred Promise with the value
+    fn resolve<'a, C: Context<'a>, T: Value>(self, cx: &mut C, v: Handle<T>);
+
+    /// Rejects a deferred Promise with the error
+    fn reject<'a, C: Context<'a>, E: Value>(self, cx: &mut C, err: Handle<E>);
+
+    /// Resolves or rejects a deferred Promise with the contents of a `Result`
+    fn complete<'a, C: Context<'a>, T: Value, E: Value>(self, cx: &mut C, result: Result<T, E>);
+}
+```
+
+Similar to [`Persistent`](https://github.com/neon-bindings/rfcs/pull/32), if a `Deferred` is not resolved or rejected, the `Promise` will leak. To help catch this and guide users towards a correct implementation, `Deferred` should `panic` on `Drop` if not used.
+
+```rust
+impl std::ops::Drop for Deferred {
+    fn drop(&mut self) {
+        panic!("JsPromise leaked. Deferred must be used.");
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None? :grin:
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
```

Comment:
We should discuss naming alternatives here (e.g. for `Deferred` as well as `cx.deferred()` vs `cx.promise()`) as we come up with them.

Reply:
Do you have any suggestions?